core/trino-main/src/main/java/io/trino/execution/scheduler/FixedCountNodeAllocator.java

```diff
@@ -1,205 +0,0 @@
-/*
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-package io.trino.execution.scheduler;
-
-import com.google.common.util.concurrent.ListenableFuture;
-import com.google.common.util.concurrent.SettableFuture;
-import io.trino.Session;
-import io.trino.connector.CatalogName;
-import io.trino.metadata.InternalNode;
-import io.trino.spi.TrinoException;
-
-import javax.annotation.concurrent.GuardedBy;
-
-import java.util.HashMap;
-import java.util.IdentityHashMap;
-import java.util.Iterator;
-import java.util.LinkedList;
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
-
-import static com.google.common.base.Preconditions.checkState;
-import static com.google.common.base.Verify.verify;
-import static com.google.common.collect.ImmutableList.toImmutableList;
-import static com.google.common.util.concurrent.Futures.immediateFailedFuture;
-import static com.google.common.util.concurrent.Futures.immediateFuture;
-import static io.trino.spi.StandardErrorCode.NO_NODES_AVAILABLE;
-import static java.util.Comparator.comparing;
-import static java.util.Objects.requireNonNull;
-
-public class FixedCountNodeAllocator
```

Comment:
Service is a singleton which you can use to get an allocator for given query.

Reply:
It could be just a single interface but with two, resource management is more natural. You have per-query object in place which you close when query completes.